File: src/main/java/org/example/parking/ParkingCar.java
package org.example.parking;

import java.time.Duration;
import java.time.LocalDateTime;

public class ParkingCar {
    private String identifier;
    private LocalDateTime enterTime;
    private LocalDateTime leaveTime;

    public ParkingCar(String identifier, LocalDateTime enterTime) {
        this.identifier = identifier;
        this.enterTime = enterTime;
    }

    public void setLeaveTime(LocalDateTime leaveTime) {
        if (leaveTime == null) {
            System.out.println("No null allowed here.");
            return;
        }
        if (!enterTime.isBefore(leaveTime)) {
            System.out.println("Invalid leave time: " + leaveTime);
            return;
        }
        this.leaveTime = leaveTime;
    }

    public Duration getDuration() {
        if (leaveTime == null) {
            throw new RuntimeException("Trying to get parking duration before car leaving.");
        }
        return Duration.between(enterTime, leaveTime);
    }
}

File: src/main/kotlin/org/example/parking/ParkingCar.kt
package org.example.parking.kotlin

import java.time.Duration
import java.time.LocalDateTime

class ParkingCar(private val identifier: String, private val enterTime: LocalDateTime) {
    var leaveTime: LocalDateTime? = null
        // NOTE: this is the syntax to declare a setter function
        // TODO: to know more details?
        // TODO: is it possible to expose only setter but not getter?
        set(newLeaveTime) {
            if (newLeaveTime == null) {
                println("No null allowed here.")
                return
            }
            if (!enterTime.isBefore(newLeaveTime)) {
                println("Invalid leave time.")
                return
            }
            // NOTE: field is the token to use to set the actual property,
            //       otherwise infinite recursive happens
            field = newLeaveTime
        }

    // TODO: null check/handling?
    fun getDuration() : Duration {
        return Duration.between(enterTime, leaveTime)
    }
}